### DISCUSSION TITLE
vi.mock in Browser mode with Nx (monorepo) hangs over

### DISCUSSION THREAD

**Discussion by dhythm** [vitest-dev/vitest#6545] [Q&A] [answered]
## Problem

Hang out the test that has `vi.mock` due to fail to register Service Workder.

I use Nx (monorepo), so all tests runs under each libraries.
The error is shown when test with `vi.mock` runs.
(vi.mock works when I use vitest browser mode without Nx.)

<img width="556" alt="image" src="https://github.com/user-attachments/assets/d7b3a603-4e76-4f8e-ba94-314593f1dd1f">

This repo reproduces the issue.
https://github.com/dhythm/vitest-browser-mode-with-nx-example

vite.config.ts in a library of Nx has `root: __dirname`.
And `root: __dirname` would be related to this issue because `vi.mock` is called when I commented out it.
(However, the test fails the other issue.)

![image](https://github.com/user-attachments/assets/f4972e0b-9949-43b6-b086-5473e4fa48a0)

Does anyone face a similar issue and have a solution to this?

**Comment by hi-ogawa** [comment, marked as answer]
I checked your reproduction and it looks like there's something wrong with `server.fs.allow`. Adding this helps getting `__vitest_msw__` work:

```ts
// vite.config.ts
  server: {
    fs: {
      allow: [path.join(__dirname, "../..")]
    }
  },
```

Another issue (partly on your code) is that you have to use `vi.hoisted` to use variables inside `vi.mock`

```ts
// [not ok]
// const mockedIsOk = vi.fn().mockReturnValue(true);
// [ok]
const mockedIsOk = vi.hoisted(() => vi.fn().mockReturnValue(true))

vi.mock('./isOk', async (importActual) => {
  // ...
  return {
    isOk: mockedIsOk,
  };
});
```

Without browser mode, the wrong code shows `ReferenceError: Cannot access 'mockedIsOk' before initialization`, but it looks like browser mode will get this stuck forever, which could be considered a bug on Vitest side.

**Reply by hi-ogawa** [reply]
On Vitest 2.1.1, It looks like the second issue doesn't exist and it shows a following error without getting stuck:

```
TypeError: Failed to fetch dynamically imported module: http://localhost:5173/src/lib/isOk.ts
```

**Reply by dhythm** [reply]
Thanks for your reply.
`server.fs.allow` solves the problem in my local environment 🙏

And I changed from `vi.mock` to `vi.hoist` too.
